### PR TITLE
EVMHost: Add tracing features needed for fuzzing.

### PR DESCRIPTION
**Pull request by bshastry**

### CODE REVIEW

File: test/EVMHost.h [hunk 1]
```diff
@@ -92,9 +87,24 @@ class EVMHost: public evmc::MockedHost
...
 
+	/// @returns the state as a string. State includes storage at and balance
+	/// of account at @param _addr and execution trace of the host post reset.
+	std::string dumpState(evmc::address _addr);
+
 private:
 	evmc::address m_currentAddress = {};
 
+	/// Records calls made via @param _message.
+	void recordCalls(evmc_message const& _message) noexcept;
```
Comment (axic) [inline test/EVMHost.h:94]: Why do we have a mixed naming style in this file?

Comment (bshastry) [reply test/EVMHost.h:94]: I changed to camel case hopefully uniformly in this PR. There are still underscored names for functions inherited from MockedHost.

**Comment by ekpyron** [comment]
@bshastry This certainly doesn't hurt, but does the fuzzer actually generate external calls? If so how does it choose addresses, s.t. not basically all of them revert? Does the interpreter deal with them?

**Comment by bshastry** [comment]
> @bshastry This certainly doesn't hurt, but does the fuzzer actually generate external calls? If so how does it choose addresses, s.t. not basically all of them revert? Does the interpreter deal with them?

The fuzzer generates external calls. It does not have any logic to choose addresses i.e., no feedback from EVMHost which basically means they are virtually useless as things stand. The (yul) interpreter does not really deal with calls (see https://github.com/ethereum/solidity/blob/b4c08b9bfe5edfe1d140a95c08269ebb989dc619/test/tools/yulInterpreter/EVMInstructionInterpreter.cpp#L304-L324).

By the way, one interesting side effect of the tracing addition in this PR (perhaps interesting is subjective) was the following "issue" found by the fuzzer

```
{
        function f(i) -> o1, o2, o3, o4
        {
                function g() -> og1, og2, og3, og4
                {
                        og1, og2, og3, og4 := f(delegatecall(0x0, 0x0, 0x0, 0x0, 0x0, 0x0))
                }
                let u, v, x, y := g()
        }
        let g1, g2, g3, g4 := f(0)
}
```

The code above produces a different "trace" (as per this PR). This is because, the optimised code runs out of EVM stack (eventually throwing stack overflow error) much later than the unoptimised code. This means the former is able to make more "delegatecall"s than the unoptimised calls, and because each delegatecall is a single line in the trace, the optimised trace looks different :-)

**Comment by bshastry** [comment]
You make a good point @ekpyron . Perhaps, we should feed in a list of valid addresses to the proto converter that may be used for call generation.

File: test/EVMHost.cpp [hunk 2]
```diff
@@ -167,11 +169,21 @@ void EVMHost::selfdestruct(const evmc::address& _addr, const evmc::address& _ben
...
 	accounts[_beneficiary].balance = balance;
 	// Record self destructs
-	recorded_selfdestructs.push_back({_addr, _beneficiary});
+	recorded_selfdestructs.push_back({_addr, _beneficiary, balance});
+}
+
+void EVMHost::recordCalls(evmc_message const& _message) noexcept
+{
+	if (recorded_calls.empty())
+		recorded_calls.reserve(max_recorded_calls);
```
Comment (ekpyron) [inline test/EVMHost.cpp]: Should this rather be ``if(!recorded_calls.capacity())``?
Also do we want to run it here every time anyways?
Seems to me that, if at all, this should be done in the constructor and probably the constructor of `MockedHost` at that... note that `.clear()` on a vector does not reset its capacity.

File: test/EVMHost.cpp [hunk 3]
```diff
@@ -784,3 +785,65 @@ StorageMap const& EVMHost::get_address_storage(evmc::address const& _addr)
 	assertThrow(account_exists(_addr), Exception, "Account does not exist.");
 	return accounts[_addr].storage;
 }
+
+void EVMHost::printCallRecords(std::ostringstream& _os) const noexcept
+{
+	auto callKind = [](evmc_call_kind _kind) -> string
```
Comment (ekpyron) [inline test/EVMHost.cpp]:
```suggestion
	static const auto callKind = [](evmc_call_kind _kind) -> string
```
?

**Review by ekpyron — COMMENTED** [review]
Some comments, but generally, I'm fine with merging this (and sort out how best to make use of it in fuzzing afterwards).
But I also wonder: since there are more and more of these tracing functions in the `EVMHost`:
All or almost all of the information that is printed by these functions is `public`, isn't it?
Might it be cleaner to move all of it into a separate `EVMHostPrinter` class or something like that?

File: test/EVMHost.cpp (hunk 2 above)
Comment (bshastry) [reply test/EVMHost.cpp]: Thank you. Fixed this.

**Comment by bshastry** [comment]
> Some comments, but generally, I'm fine with merging this (and sort out how best to make use of it in fuzzing afterwards).
> But I also wonder: since there are more and more of these tracing functions in the `EVMHost`:
> All or almost all of the information that is printed by these functions is `public`, isn't it?
> Might it be cleaner to move all of it into a separate `EVMHostPrinter` class or something like that?

Thanks, I now moved the "printing" code to a separate class.

File: test/EVMHost.cpp [hunk 4]
```diff
@@ -761,18 +774,7 @@ evmc::result EVMHost::resultWithGas(
...
-	{
-		_os << "Address: " << convertFromEVMC(addr) << endl;
-		for (auto const& [slot, value]: get_address_storage(addr))
-			if (get_storage(addr, slot))
-				_os << convertFromEVMC(slot) << ": " << convertFromEVMC(value.value) << endl;
-	}
-}
-
-void EVMHost::print_storage_at(evmc::address const& _addr, ostringstream& _os)
+void EVMHost::printStorageAt(evmc::address const& _addr, ostringstream& _os)
```
Comment (ekpyron) [inline test/EVMHost.cpp]:
```suggestion
void EVMHostPrinter
::printStorageAt(evmc::address const& _addr, ostringstream& _os)
```

**Review by ekpyron — APPROVED** [review]
Looks fine to me now.

**Pull request merged**